File: app/src/main/java/projects/csce/evence/view/ui/MainActivity.java
package projects.csce.evence.view.ui;

import android.content.Intent;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.ActionBarDrawerToggle;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.google.android.gms.auth.api.signin.GoogleSignIn;
import com.google.android.gms.auth.api.signin.GoogleSignInAccount;
import com.google.android.gms.auth.api.signin.GoogleSignInClient;
import com.google.android.gms.common.api.ApiException;
import com.google.android.gms.tasks.Task;

import javax.inject.Inject;

import daniel.perez.core.BaseActivity;
import daniel.perez.core.DialogStarter;
import daniel.perez.core.StartActivity;
import daniel.perez.core.adapter.CardsAdapter;
import daniel.perez.core.di.ViewModelFactory;
import daniel.perez.core.service.FileManager;
import daniel.perez.core.service.qr.QrBitmapGenerator;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import projects.csce.evence.BaseApplication;
import projects.csce.evence.R;
import projects.csce.evence.databinding.ActivityMainBinding;
import projects.csce.evence.utils.Utils;
import projects.csce.evence.viewmodel.MainViewModel;
import timber.log.Timber;

public class MainActivity extends BaseActivity
{
    private int RC_SIGN_IN = 1;
    private MainViewModel viewModel;
    private ActivityMainBinding binding;
    private CardsAdapter eventsAdapter;
    private CompositeDisposable disposables = new CompositeDisposable();

    @Inject FileManager fileManager;
    @Inject GoogleSignInClient signInClient;
    @Inject ViewModelFactory factory;
    @Inject QrBitmapGenerator generator;
    @Inject DialogStarter dialogStarter;
    @Inject StartActivity startActivity;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        setTheme(R.style.AppTheme);
        Utils.getAppComponent(this).inject(this);
        super.onCreate(savedInstanceState);
        binding = ActivityMainBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        binding.qrBtn.setOnClickListener( unit -> startQrActivity());

        //apply custom toolbar
        setSupportActionBar(binding.toolbarMain);


        viewModel = ViewModelProviders.of(this, factory).get(MainViewModel.class);
        eventsAdapter = new CardsAdapter(this);
        handleRecyclerView();

        //binding.loginBtn.setOnClickListener(view -> signIn());

        //handle drawer
        ActionBarDrawerToggle actionBarDrawerToggle = new ActionBarDrawerToggle(this, binding.drawerMain, binding.toolbarMain, R.string.app_name, R.string.app_name);
        binding.drawerMain.closeDrawer(binding.includedDrawer.navigationDrawer);
        binding.drawerMain.addDrawerListener(actionBarDrawerToggle);
        actionBarDrawerToggle.syncState();

        setupSubscriptions();
    }

    private void setupSubscriptions()
    {
        disposables.add(viewModel.liveFiles()
                .doOnSubscribe(event -> fileManager.notifyIcals())
                .subscribe(files -> {
                    Timber.i("Received Files");
                    eventsAdapter.setData(files);
                }));

        disposables.add(eventsAdapter.clicks()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(viewCalendarData -> dialogStarter.startQrDialog(this, viewCalendarData)));

        Timber.i("Set up subscriptions");
    }

    @Override
    protected void onResume() {
        super.onResume();
        fileManager.notifyIcals();
    }

    @Override
    protected void onDestroy()
    {
        super.onDestroy();
        disposables.dispose();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        if (item.getItemId() == R.id.qr_camera_btn)
        {
            startQrReaderActivity();
        }
        return super.onOptionsItemSelected(item);
    }

    private void handleRecyclerView() {
        binding.eventsRecyclerView.setAdapter(eventsAdapter);
        binding.eventsRecyclerView.setLayoutManager(new LinearLayoutManager(getBaseContext()));
        disposables.add(viewModel.liveFiles().subscribe( events -> {
            Timber.i(Integer.toString(events.size()));
            eventsAdapter.onChanged(events);
            if (events.size() == 0)
                binding.emptyTextview.setVisibility(View.VISIBLE);
            else
                binding.emptyTextview.setVisibility(View.GONE);
        }));
    }

    public void startQrReaderActivity()
    {
        startActivity.startQrReader(this);
    }

    public void startQrActivity() {
        startActivity.startGenerateQr(this);
    }

    public void startSecondActivity() {
        startActivity.startSecondActivity(this);
    }

    public void startSettingsActivity(View view) {
        binding.drawerMain.closeDrawer(binding.includedDrawer.navigationDrawer);
        startActivity.startSettingsActivity(this);
    }

    public void startShareAppActivity(View view) {
        binding.drawerMain.closeDrawer(binding.includedDrawer.navigationDrawer);
        startActivity.startShareAppActivity(this);
    }

    public void startAboutActivity(View view){
        binding.drawerMain.closeDrawer(binding.includedDrawer.navigationDrawer);
        startActivity.startAboutActivity(this);
    }

    public void signIn() {
        Intent signInIntent = signInClient.getSignInIntent();
        startActivityForResult(signInIntent, RC_SIGN_IN);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        // Result returned from launching the Intent from GoogleSignInClient.getSignInIntent(...);
        Timber.d(Integer.toString(requestCode));
        if (requestCode == RC_SIGN_IN) {
            // The Task returned from this call is always completed, no need to attach
            // a listener.
            Task<GoogleSignInAccount> task = GoogleSignIn.getSignedInAccountFromIntent(data);
            handleSignInResult(task);
        }
    }

    private void handleSignInResult(Task<GoogleSignInAccount> completedTask) {
        try {
            ((BaseApplication) getApplication()).setAccount(completedTask.getResult(ApiException.class));
            // Signed in successfully, show authenticated UI.
            startSecondActivity();

        } catch (ApiException e) {
            Timber.tag("SigninAttempt").w(Integer.toString(e.getStatusCode()));
            e.printStackTrace();
            Toast.makeText(getApplicationContext(), "Sign-In Failed", Toast.LENGTH_SHORT).show();
        }
    }
}

File: app/src/main/java/projects/csce/evence/view/ui/MainActivity.kt
package projects.csce.evence.view.ui

import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.PopupMenu
import android.widget.Switch
import androidx.appcompat.app.ActionBarDrawerToggle
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import coil.ImageLoader
import com.jakewharton.rxbinding4.widget.textChanges
import dagger.hilt.android.AndroidEntryPoint
import daniel.perez.core.*
import daniel.perez.core.RequestCodes.REQUEST_SAF
import daniel.perez.core.adapter.CardsAdapter
import daniel.perez.core.model.ViewEvent
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import io.reactivex.rxjava3.disposables.Disposable
import io.reactivex.rxjava3.functions.Consumer
import projects.csce.evence.R
import projects.csce.evence.databinding.ActivityMainBinding
import projects.csce.evence.service.model.SharedPref
import projects.csce.evence.viewmodel.MainViewModel
import projects.csce.evence.viewmodel.SortState
import timber.log.Timber
import java.util.concurrent.TimeUnit
import javax.inject.Inject

@AndroidEntryPoint
class MainActivity : BaseActivity(), AdapterView.OnItemSelectedListener
{
    private lateinit var currentEvent: ViewEvent
    private lateinit var viewModel: MainViewModel
    private lateinit var binding: ActivityMainBinding
    private lateinit var eventsAdapter: CardsAdapter

    @Inject lateinit var dialogStarter: DialogStarter
    @Inject lateinit var activityStarter: ActivityStarter
    @Inject lateinit var sharedPref: SharedPref
    @Inject lateinit var imageLoader: ImageLoader
    @Inject lateinit var activityResultActions: ActivityResultActions

    override fun onCreate(savedInstanceState: Bundle?)
    {
        super.onCreate(savedInstanceState)
        viewModel = ViewModelProvider(this).get(MainViewModel::class.java)
        setTheme(R.style.AppTheme)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)
        viewSetup()
        setupSubscriptions()
    }

    private fun viewSetup()
    {
        binding.qrBtn.setOnClickListener {
//            activityStarter.startNewEventActivity(this)
            activityStarter.startGenerateQr(this)
        }

        //apply custom toolbar
        setSupportActionBar(binding.toolbarMain)

        eventsAdapter = CardsAdapter()

        //handle drawer
        val actionBarDrawerToggle = ActionBarDrawerToggle(this, binding.drawerMain, binding.toolbarMain, R.string.app_name, R.string.app_name)
        binding.drawerMain.closeDrawer(binding.includedDrawer.navigationDrawer)
        binding.drawerMain.addDrawerListener(actionBarDrawerToggle)
        actionBarDrawerToggle.syncState()

        // RecyclerView
        binding.eventsRecyclerView.adapter = eventsAdapter
        binding.eventsRecyclerView.layoutManager = LinearLayoutManager(baseContext)

    }

    private fun setupSubscriptions()
    {
        disposables += viewModel.events()
                .observeOn( AndroidSchedulers.mainThread() )
                .subscribe{ events: List<ViewEvent> ->
                    eventsAdapter.setData( events )
                    if (events.isEmpty())
                    {
                        binding.emptyTextview.visibility = View.VISIBLE
                    }
                    else
                    {
                        binding.emptyTextview.visibility = View.GONE
                    }
                }

        disposables += binding.searchEditText.textChanges()
                .debounce(100, TimeUnit.MILLISECONDS)
                .map { it.toString() }
                .distinctUntilChanged()
                .subscribe {
                    Timber.i("""Text Change Subscribe: '$it'""")
                    viewModel.newSearch( it.toString() )
                }

        disposables += eventsAdapter.clicks()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe {
                    currentEvent = it
                    dialogStarter.startQrDialog(this, it)
                }

        disposables += sharedPref.getUiPref()
                .doOnSubscribe { pref: Disposable? -> sharedPref.notifyUiPref() }
                .subscribe { eventsAdapter.updateUiFormat(it) }
    }

    override fun onResume()
    {
        super.onResume()
        sharedPref.notifyUiPref()
    }

    override fun onStop()
    {
        super.onStop()
        binding.searchEditText.setText("")
    }

    override fun onDestroy()
    {
        super.onDestroy()
        disposables.dispose()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean
    {
        menuInflater.inflate(R.menu.menu_main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean
    {
        if (item.itemId == R.id.qr_camera_btn) {
            activityStarter.startQrReader(this)
        } else if (item.itemId == R.id.sort_btn) {
            handleSort()
        }
        return super.onOptionsItemSelected(item)
    }

    fun handleSort() {
        val popupMenu = PopupMenu(this, findViewById(R.id.sort_btn))
        popupMenu.menuInflater.inflate(R.menu.menu_sort, popupMenu.menu)
        popupMenu.setOnMenuItemClickListener { menuItem ->
            when (menuItem.itemId) {
                R.id.sort_by_creation ->
                {
                    viewModel.setSortState(SortState.RECENTLY_CREATED)
                }
                R.id.sort_by_date ->
                {
                    viewModel.setSortState(SortState.BY_DATE_ASCENDING)
                }
            }
            true
        }
        popupMenu.show()
    }

    fun startSettingsActivity(view: View?)
    {
        binding.drawerMain.closeDrawer(binding.includedDrawer.navigationDrawer)
        activityStarter.startSettingsActivity(this)
    }

    fun startShareAppActivity(view: View?)
    {
        binding.drawerMain.closeDrawer(binding.includedDrawer.navigationDrawer)
        activityStarter.startShareAppActivity(this)
    }

    fun startAboutActivity(view: View?)
    {
        binding.drawerMain.closeDrawer(binding.includedDrawer.navigationDrawer)
        activityStarter.startAboutActivity(this)
    }

    public override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?)
    {
        super.onActivityResult(requestCode, resultCode, data)

        Timber.i("${RequestCodes.map(requestCode)}, requestCode: $requestCode")
        if(resultCode == RESULT_OK)
        {
            Timber.i("RESULT_OK")
            if (requestCode == REQUEST_SAF)
            {
                if(data != null)
                {
                    disposables += activityResultActions.actionCreateDocumentEvent(this, currentEvent, data)
                            .subscribe(resultHandler)
                }
            }
        }
    }

    private val resultHandler = Consumer<ActionResult> {
        when(it)
        {
            ActionResult.Success -> toastShort("Wrote File Successfully")
            is ActionResult.Failure ->
            {
                Timber.e(it.t)
                toastShort("Error writing file")
            }
            ActionResult.InTransit ->
            {
                Timber.i("Writing file...")
            }
        }
    }

    override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {

    }

    override fun onNothingSelected(parent: AdapterView<*>?) {
        TODO("Not yet implemented")
    }
}